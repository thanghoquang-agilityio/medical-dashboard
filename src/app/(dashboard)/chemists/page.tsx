const ChemistPage = () => (
  <h1 className="text-lg font-semibold">
    Welcome to Chemist page: test commit
  </h1>
);

export default ChemistPage;
