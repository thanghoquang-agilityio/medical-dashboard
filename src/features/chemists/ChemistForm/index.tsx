'use client';

import {
  ChangeEvent,
  memo,
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react';
import { Controller, SubmitHandler, useForm } from 'react-hook-form';

// Components
import { Button, ImageUpload, Input, Select, Text } from '@/components/ui';
import { EyeIcon, EyeSlashIcon, NoteIcon, StarIcon } from '@/icons';
import { Textarea } from '@nextui-org/react';

// Types
import {
  UserModel,
  ChemistFormData,
  UserPayload,
  RolePermission,
  ROLE,
  STATUS_TYPE,
  Option,
} from '@/types';
import { ERROR_MESSAGE, SUCCESS_MESSAGE } from '@/constants';

// Helper
import { clearErrorOnChange, getRoleIdByName } from '@/utils';
import { useToast } from '@/context/toast';
import { addUserToChemists } from '@/actions/auth';
import {
  addUser,
  getUserRoles,
  updatePublishUser,
  updateUser,
  uploadImage,
} from '@/services';

// Rules
import { CHEMIST_FORM_VALIDATION } from './rule';

export type ChemistFormProps = {
  specialtyOptions: Option[];
  id?: string;
  data?: UserModel;
  onClose: () => void;
};

const ChemistForm = memo(
  ({ id = '', data, specialtyOptions, onClose }: ChemistFormProps) => {
    const {
      username = '',
      email = '',
      password = '',
      avatar,
      description = '',
      rating = 0,
      tasks = 0,
      reviews = 0,
      specialtyId,
    } = data || {};

    const { id: avatarId, attributes: attributesAvatar } = avatar?.data || {};
    const { url = '' } = attributesAvatar || {};

    const { id: specialty } = specialtyId?.data || {};

    const [imageUpload, setImageUpload] = useState<string>('');
    const [formImage, setFormImage] = useState<FormData | null>(null);
    const [imageRemote, setImageRemote] = useState<string>(url);

    const [error, setError] = useState('');
    const [isPending, setIsPending] = useState(false);
    const [roles, getRoles] = useState<RolePermission[]>([]);

    const openToast = useToast();

    const {
      control,
      handleSubmit,
      setValue,
      getValues,
      clearErrors,
      setError: setFormError,
      formState: { errors, isValid, isLoading, isDirty },
    } = useForm<ChemistFormData>({
      mode: 'onBlur',
      reValidateMode: 'onBlur',
      defaultValues: {
        avatar: url,
        username,
        password,
        confirmPassWord: password,
        email,
        description,
        specialtyId: String(specialty),
      },
    });

    useEffect(() => {
      const fetchUserRoles = async () => {
        const { roles, error } = await getUserRoles();
        if (error) throw error;
        getRoles(roles);
      };

      fetchUserRoles();
    }, []);

    const isEdit = !!data;

    // Handle show hide password
    const [isShowPassword, setIsShowPassword] = useState<boolean>(false);
    const [isShowConfirmPassword, setIsShowConfirmPassword] =
      useState<boolean>(false);
    const handleToggleVisiblePassword = useCallback(
      () => setIsShowPassword((prev) => !prev),
      [],
    );
    const handleToggleShowConfirmPassword = useCallback(
      () => setIsShowConfirmPassword((prev) => !prev),
      [],
    );

    // Handle input file changes
    const handleUpload = (event: React.ChangeEvent<HTMLInputElement>) => {
      const files = event.target.files;
      if (files && files[0]) {
        const image = files[0];
        const formData = new FormData();
        formData.append('files', image);

        setFormImage(formData);
        setImageUpload(URL.createObjectURL(image));
        setValue('avatar', URL.createObjectURL(image));
      }
    };

    // Handle remove upload image
    const hiddenFileInput = useRef<HTMLInputElement>(null);
    const handleRemoveImage = () => {
      setImageUpload('');
      setFormImage(null);
      setValue('avatar', '');
      setImageRemote('');

      if (hiddenFileInput.current) {
        hiddenFileInput.current.value = '';
      }
    };

    const handleClick = useCallback(() => {
      hiddenFileInput.current?.click();
    }, []);

    const handleInputChange = useCallback(
      (name: keyof ChemistFormData, onChange: (value: string) => void) => {
        return (e: ChangeEvent<HTMLInputElement>) => {
          onChange(e.target.value);
          // Clear error message on change
          clearErrorOnChange(name, errors, clearErrors);
        };
      },
      [clearErrors, errors],
    );

    const handleError = useCallback(
      (error: string) => {
        setError(error);
        openToast({
          message: ERROR_MESSAGE.CREATE('chemist'),
          type: STATUS_TYPE.ERROR,
        });
        setIsPending(false);
        return;
      },
      [openToast],
    );

    // Handle submit form data to create chemist
    const handleSubmitForm: SubmitHandler<ChemistFormData> = useCallback(
      async (formData) => {
        setError('');
        setIsPending(true);

        const { avatar, username, email, password, specialtyId, description } =
          formData;

        const avatarUpload = formImage
          ? await uploadImage(formImage)
          : undefined;

        const payload: UserPayload = {
          username,
          email,
          description,
          specialtyId: Number(specialtyId),
          role: Number(getRoleIdByName(roles, ROLE.NORMAL_USER)),
          ...(!isEdit && { password }),
          ...(isEdit
            ? avatarUpload
              ? { avatar: Number(avatarUpload?.image?.[0]?.id) }
              : { avatar: avatar && !avatarUpload ? Number(avatarId) : null }
            : avatar && { avatar: Number(avatarUpload?.image?.[0]?.id) }),
        };

        if (isEdit) {
          const { error } = await updateUser(id, payload);

          if (error) {
            handleError(error);
            return;
          }
        } else {
          const { user, error } = await addUser(payload);

          if (error) {
            handleError(error);
            return;
          }

          if (user) {
            const { id = '' } = user;

            const updateUserResult = await updatePublishUser(id);
            if (updateUserResult.error) {
              handleError(updateUserResult.error);
              return;
            }

            const addUserResult = await addUserToChemists({
              users_permissions_user: String(id),
            });
            if (addUserResult.error) {
              handleError(addUserResult.error);
              return;
            }

            openToast({
              message: isEdit
                ? SUCCESS_MESSAGE.UPDATE('chemist')
                : SUCCESS_MESSAGE.CREATE('chemist'),
              type: STATUS_TYPE.SUCCESS,
            });
          }
        }
        setIsPending(false);
        onClose?.();
      },
      [formImage, roles, isEdit, avatarId, onClose, id, handleError, openToast],
    );

    return (
      <form
        onSubmit={handleSubmit(handleSubmitForm)}
        className="p-4 flex flex-col"
      >
        <Text variant="title" size="xl">
          {isEdit ? 'Update chemist' : 'Create chemist'}
        </Text>

        {/* Avatar */}
        <Controller
          control={control}
          name="avatar"
          render={({ field }) => (
            <ImageUpload
              {...field}
              ref={hiddenFileInput}
              src={imageRemote}
              srcUpload={imageUpload}
              onRemoveImage={handleRemoveImage}
              onUploadImage={handleUpload}
              onClick={handleClick}
            />
          )}
        />

        {/* Info section */}
        {isEdit && (
          <div className="flex justify-between self-center gap-6 mt-4">
            <div className="flex items-center gap-1 sm:gap-2">
              <NoteIcon customClass="w-6 h-6" />
              <Text size="sm" variant="title" customClass="font-medium">
                {tasks} Tasks
              </Text>
            </div>

            <div className="flex items-center gap-1 sm:gap-2">
              <StarIcon customClass="w-6 h-6 text-light-orange" />
              <Text size="sm" variant="title" customClass="font-medium">
                {rating} ({reviews} Reviews)
              </Text>
            </div>
          </div>
        )}

        {/* Username */}
        <Controller
          name="username"
          control={control}
          render={({ field: { name, ...rest }, fieldState: { error } }) => (
            <Input
              {...rest}
              isRequired
              label="Username"
              labelPlacement="outside"
              name={name}
              placeholder="Please enter username"
              type="text"
              size="sm"
              isDisabled={isLoading || isEdit}
              isInvalid={!!error?.message}
              errorMessage={error?.message}
            />
          )}
          rules={CHEMIST_FORM_VALIDATION.USERNAME}
        />

        {/* Email */}
        <Controller
          name="email"
          control={control}
          render={({ field: { name, ...rest }, fieldState: { error } }) => (
            <Input
              {...rest}
              label="Email"
              isRequired
              labelPlacement="outside"
              name={name}
              placeholder="Please enter email"
              type="text"
              size="sm"
              isDisabled={isLoading || isEdit}
              isInvalid={!!error?.message}
              errorMessage={error?.message}
            />
          )}
          rules={CHEMIST_FORM_VALIDATION.EMAIL}
        />

        {!isEdit && (
          <>
            {/* Password */}
            <Controller
              name="password"
              control={control}
              render={({
                field: { onChange, name, ...rest },
                fieldState: { error },
              }) => (
                <Input
                  {...rest}
                  name={name}
                  label="Password"
                  labelPlacement="outside"
                  size="sm"
                  type={isShowPassword ? 'text' : 'password'}
                  placeholder="Please enter password"
                  endContent={
                    <Button
                      aria-label="visible password"
                      onClick={handleToggleVisiblePassword}
                      isIconOnly
                      className="p-0 min-w-5 h-5 text-primary-200"
                    >
                      {isShowPassword ? <EyeIcon /> : <EyeSlashIcon />}
                    </Button>
                  }
                  isDisabled={isLoading}
                  isInvalid={!!error?.message}
                  errorMessage={error?.message}
                  onChange={handleInputChange(name, onChange)}
                />
              )}
              rules={CHEMIST_FORM_VALIDATION.PASSWORD(
                getValues,
                setFormError,
                clearErrors,
                isEdit,
              )}
            />

            {/* Confirm Password */}
            <Controller
              name="confirmPassWord"
              control={control}
              render={({
                field: { onChange, name, ...rest },
                fieldState: { error },
              }) => (
                <Input
                  {...rest}
                  size="sm"
                  label="Confirm Password"
                  labelPlacement="outside"
                  type={isShowConfirmPassword ? 'text' : 'password'}
                  placeholder="Please enter confirm password"
                  endContent={
                    <Button
                      aria-label="visible confirm password"
                      onClick={handleToggleShowConfirmPassword}
                      isIconOnly
                      className="p-0 min-w-5 h-5 text-primary-200"
                    >
                      {isShowConfirmPassword ? <EyeIcon /> : <EyeSlashIcon />}
                    </Button>
                  }
                  isDisabled={isLoading}
                  isInvalid={!!error?.message}
                  errorMessage={error?.message}
                  onChange={handleInputChange(name, onChange)}
                />
              )}
              rules={CHEMIST_FORM_VALIDATION.CONFIRM_PASSWORD(
                getValues,
                isEdit,
              )}
            />
          </>
        )}

        {/* Specialty */}
        <Controller
          control={control}
          name="specialtyId"
          render={({
            field: { name, value, onChange, ...rest },
            fieldState: { error },
          }) => (
            <Select
              {...rest}
              name={name}
              value={value}
              label="Specialty"
              placeholder="Select specialty"
              labelPlacement="outside"
              variant="bordered"
              classNames={{
                mainWrapper: 'h-16',
                trigger: 'h-[auto] py-3 max-h-10',
                errorMessage: 'text-danger text-xs ml-2',
                label: 'top-5 text-sm text-foreground',
                value: 'text-sm text-primary-100',
              }}
              isDisabled={isLoading}
              defaultSelectedKeys={[value]}
              options={specialtyOptions}
              isInvalid={!!error?.message}
              errorMessage={error?.message}
              onChange={onChange}
            />
          )}
          rules={CHEMIST_FORM_VALIDATION.SPECIALTY}
        />

        {/* Description */}
        <Controller
          name="description"
          control={control}
          render={({
            field: { name, value, ...rest },
            fieldState: { error },
          }) => (
            <Textarea
              {...rest}
              label="Description"
              labelPlacement="outside"
              name={name}
              value={value}
              size="sm"
              classNames={{
                inputWrapper:
                  'border-text-foreground border-1 data-[focus=true]:border-secondary-300 group-data-[focus-visible=true]:ring-0',
                errorMessage: 'text-danger text-xs ml-2',
                label: 'text-sm',
              }}
              isDisabled={isLoading}
              placeholder="Please enter description"
              isInvalid={!!error?.message}
              errorMessage={error?.message}
            />
          )}
          rules={CHEMIST_FORM_VALIDATION.DESCRIPTION}
        />

        <div className="h-[78px] flex flex-col justify-end">
          {error && (
            <Text variant="error" size="sm" customClass="py-2">
              {error}
            </Text>
          )}
          <div className="w-full gap-2 flex justify-end">
            <Button
              onClick={onClose}
              variant="outline"
              color="outline"
              className="font-medium"
            >
              Cancel
            </Button>
            <Button
              isDisabled={!isValid || !isDirty || isPending}
              isLoading={isLoading || isPending}
              type="submit"
            >
              Submit
            </Button>
          </div>
        </div>
      </form>
    );
  },
);

export default ChemistForm;
ChemistForm.displayName = 'ChemistForm';
