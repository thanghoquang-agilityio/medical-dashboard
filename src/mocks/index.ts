export * from './appointments';
export * from './auth';
export * from './chemists';
export * from './common';
export * from './notification';
export * from './profile';
export * from './table';
export * from './user';
