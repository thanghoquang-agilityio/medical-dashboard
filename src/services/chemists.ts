'use server';
import { revalidateTag } from 'next/cache';

// Types
import {
  ChemistDataResponse,
  ChemistPayload,
  ChemistResponse,
  ChemistsDataResponse,
  ChemistsResponse,
  ErrorResponse,
  FetchDataProps,
} from '@/types';

// Constants
import {
  API_ENDPOINT,
  EXCEPTION_ERROR_MESSAGE,
  HOST_DOMAIN,
  ROUTE_ENDPOINT,
} from '@/constants';
import { auth } from '@/config/auth';

export const addUserToChemists = async (
  payload: ChemistPayload,
): Promise<ChemistDataResponse> => {
  try {
    const response = await fetch(
      `${HOST_DOMAIN}/${ROUTE_ENDPOINT.CHEMISTS.ADD_TO_CHEMISTS}`,
      {
        method: 'POST',
        body: JSON.stringify(payload),
      },
    );
    const {
      data,
      error,
    }: {
      data: ChemistResponse;
      error?: string;
    } = await response.json();

    if (error) {
      return {
        chemist: null,
        error: (JSON.parse(error) as ErrorResponse).error.message,
      };
    }

    revalidateTag(API_ENDPOINT.CHEMISTS);

    return { chemist: data, error: null };
  } catch (error) {
    const errorMessage =
      error instanceof Error
        ? error.message
        : EXCEPTION_ERROR_MESSAGE.ADD('user to chemists');

    return { chemist: null, error: errorMessage };
  }
};

export const getChemists = async ({
  searchParams = new URLSearchParams(),
  options = { next: { tags: [API_ENDPOINT.CHEMISTS] } },
}: FetchDataProps): ChemistsDataResponse => {
  try {
    const { token = '' } = (await auth())?.user || {};

    const url = decodeURIComponent(
      `${HOST_DOMAIN}/${ROUTE_ENDPOINT.CHEMISTS.GET_CHEMISTS}?${searchParams.toString()}`,
    );

    const response = await fetch(url, {
      ...options,
      headers: {
        Authorization: `Bearer ${token}`,
      },
      next: {
        ...options.next,
        revalidate: 3600,
      },
    });

    const { data, meta, error }: ChemistsResponse & { error?: string } =
      await response.json();

    if (error) {
      const errorResponse = JSON.parse(error) as ErrorResponse;
      return { chemists: [], error: errorResponse.error.message };
    }

    return {
      chemists: data,
      ...meta,
      error: null,
    };
  } catch (error) {
    const errorMessage =
      error instanceof Error
        ? error.message
        : EXCEPTION_ERROR_MESSAGE.GET('chemists');

    return { chemists: [], error: errorMessage };
  }
};
