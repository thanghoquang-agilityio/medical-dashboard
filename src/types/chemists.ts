import {
  APIRelatedResponse,
  APIResponse,
  MetaResponse,
  UserModel,
} from '@/types';

export type ChemistModel = {
  users_permissions_user: APIRelatedResponse<APIResponse<UserModel>>;
};

export type ChemistsDataResponse = Promise<
  { chemists: ChemistResponse[]; error: string | null } & MetaResponse
>;

export type ChemistsResponse = {
  data: ChemistResponse[];
  meta: MetaResponse;
};

export type ChemistPayload = {
  users_permissions_user: string;
};

export type ChemistResponse = {
  error: string | null;
};
