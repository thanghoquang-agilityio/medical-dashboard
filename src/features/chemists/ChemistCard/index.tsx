'use client';
import { lazy, useCallback, useState } from 'react';

// Components
import {
  Card,
  CardBody,
  CardFooter,
  CardHeader,
  useDisclosure,
} from '@nextui-org/react';
import { Avatar, Button, Text } from '@/components/ui';
import { CloseIcon, NoteIcon, StarIcon } from '@/icons';
const ConfirmModal = lazy(() => import('@/components/ui/ConfirmModal'));

// Types
import { STATUS_TYPE, UserModel } from '@/types';
import { API_IMAGE_URL, ERROR_MESSAGE, SUCCESS_MESSAGE } from '@/constants';

// Utils
import { cn } from '@/utils';
import {
  updateUnpublishAppointment,
  updateUnpublishNotification,
  updateUnpublishUser,
} from '@/services';
import { useToast } from '@/context/toast';

export interface ChemistCardProps {
  id: string;
  data: UserModel;
  isAdmin: boolean;
  onEdit?: ({ data, id }: { data: UserModel; id: string }) => void;
}

const ChemistCard = ({ id, data, isAdmin, onEdit }: ChemistCardProps) => {
  const {
    username = '',
    description = '',
    rating = 0,
    reviews,
    tasks,
    avatar,
    specialtyId,
  } = data || {};

  const { data: dataAvatar } = avatar || {};
  const { attributes: attributesAvatar } = dataAvatar || {};
  const { url = '' } = attributesAvatar || {};

  const { data: dataSpecialty } = specialtyId || {};
  const { attributes: attributesSpecialty } = dataSpecialty || {};
  const { name: specialty = '' } = attributesSpecialty || {};

  const openToast = useToast();

  const {
    isOpen: isOpenConfirm,
    onClose: onClosConfirm,
    onOpen: onOpenConfirm,
  } = useDisclosure();

  const [isLoading, setIsLoading] = useState(false);

  const handleOpenConfirmModal = useCallback(() => {
    onOpenConfirm();
  }, [onOpenConfirm]);

  const handleDelete = useCallback(async () => {
    setIsLoading(true);
    const { error: errorUser } = await updateUnpublishUser(id);
    if (errorUser) {
      openToast({
        message: ERROR_MESSAGE.DELETE('chemist'),
        type: STATUS_TYPE.ERROR,
      });
      setIsLoading(false);
      return;
    }

    openToast({
      message: SUCCESS_MESSAGE.DELETE('chemist'),
      type: STATUS_TYPE.SUCCESS,
    });

    setIsLoading(false);
    onClosConfirm();

    await updateUnpublishAppointment(id);
    await updateUnpublishNotification(id);
  }, [id, onClosConfirm, openToast]);

  const handleEdit = useCallback(() => {
    onEdit?.({ data, id });
  }, [data, id, onEdit]);

  return (
    <div
      className={cn(
        `min-w-[300px] w-full h-[228px] ${isAdmin ? 'cursor-pointer' : 'cursor-default'}`,
      )}
      onClick={handleEdit}
    >
      <Card className="bg-background-200 w-full h-full p-5 sm:p-6 gap-6 overflow-visible">
        {isAdmin && (
          <Button
            aria-label="close"
            isIconOnly
            size="tiny"
            color="red"
            className="absolute top-[-8px] right-[-8px] min-w-6"
            onClick={handleOpenConfirmModal}
          >
            <CloseIcon />
          </Button>
        )}
        <CardHeader className="flex justify-between p-0">
          <div className="flex items-center gap-2">
            <Avatar src={`${API_IMAGE_URL}${url}`} size="lg" />
            <div className="flex flex-col gap-1">
              <Text size="md" variant="title">
                {username}
              </Text>
              {specialty && (
                <Text size="xs" variant="description" customClass="font-normal">
                  {specialty}
                </Text>
              )}
            </div>
          </div>
          <Button
            color="default"
            className="text-green text-lg font-medium p-0 min-w-10"
          >
            Book
          </Button>
        </CardHeader>
        <CardBody className="p-0">
          <Text
            size="sm"
            type="wrap"
            variant="description"
            customClass="line-clamp-2"
          >
            {description || 'No description'}
          </Text>
        </CardBody>
        <CardFooter className="p-0 flex justify-between">
          <div className="flex items-center gap-1 sm:gap-2">
            <NoteIcon customClass="w-6 h-6" />
            <Text size="sm" variant="title" customClass="font-medium">
              {tasks} Task
            </Text>
          </div>

          <div className="flex items-center gap-1 sm:gap-2">
            <StarIcon customClass="w-6 h-6 text-light-orange" />
            <Text size="sm" variant="title" customClass="font-medium">
              {rating} ({reviews} Reviews)
            </Text>
          </div>
        </CardFooter>
      </Card>

      <ConfirmModal
        title="Confirmation"
        subTitle={`Do you want to delete this chemist?`}
        isOpen={isOpenConfirm}
        isLoading={isLoading}
        onClose={onClosConfirm}
        onAction={handleDelete}
      />
    </div>
  );
};

export default ChemistCard;
