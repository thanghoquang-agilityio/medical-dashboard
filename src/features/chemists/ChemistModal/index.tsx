'use client';

import { memo } from 'react';

// Types
import { AppointmentModel } from '@/types';

// Components
import { BaseModal } from '@/components/ui/BaseModal';
import ChemistForm from '../ChemistForm';

// Utils
import { USER_OPTIONS } from '@/mocks/user';

export type ChemistModalProps = {
  isOpen: boolean;
  onClose: () => void;
  data?: AppointmentModel;
  id?: string;
};

const ChemistModal = memo(({ isOpen, onClose, id }: ChemistModalProps) => {
  return (
    <BaseModal isOpen={isOpen} onClose={onClose} placement="center" size="2xl">
      {/* TODO: Update later */}
      <ChemistForm data={USER_OPTIONS[0]} id={id} />
    </BaseModal>
  );
});

export default ChemistModal;
ChemistModal.displayName = 'ChemistModal';
