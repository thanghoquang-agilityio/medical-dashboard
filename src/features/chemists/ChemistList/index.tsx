'use client';

import { useDisclosure } from '@nextui-org/react';
import { usePathname, useRouter, useSearchParams } from 'next/navigation';
import {
  Key,
  lazy,
  memo,
  Suspense,
  useCallback,
  useMemo,
  useState,
  useTransition,
} from 'react';

// Types
import {
  APIResponse,
  ChemistModel,
  MetaResponse,
  ROLE,
  SpecialtyModel,
  UserModel,
} from '@/types';

// Components
import { Button, InputSearch, MenuDropdown, Text } from '@/components/ui';
import { CategoryIcon } from '@/icons';
import ChemistCard from '../ChemistCard';
import ChemistModal from '../ChemistModal';
import { ChemistListSkeleton } from './ChemistSkeleton';

// Utils
import { formatString, transformSpecialties } from '@/utils';

// Constants
import { PAGE_DEFAULT, RESULT_NOT_FOUND } from '@/constants';

const Pagination = lazy(() => import('@/components/ui/Pagination'));

export type ChemistListProps = {
  chemists: Array<APIResponse<ChemistModel>>;
  specialties: Array<APIResponse<SpecialtyModel>>;
  role: string;
  defaultSpecialty?: string;
} & MetaResponse;

const ChemistList = memo(
  ({
    chemists,
    role,
    pagination,
    defaultSpecialty,
    specialties,
  }: ChemistListProps) => {
    const { page = PAGE_DEFAULT, pageCount = PAGE_DEFAULT } = pagination ?? {};

    const [chemist, setChemist] = useState<UserModel>();
    const [chemistId, setChemistId] = useState<string>('');

    const [isPending, startTransition] = useTransition();
    const { isOpen, onOpen, onClose } = useDisclosure();

    const searchParams = useSearchParams() ?? '';
    const pathname = usePathname() ?? '';
    const { replace } = useRouter();

    const specialtyOptions = transformSpecialties(specialties);
    const specialtyOptionsForm = specialtyOptions.slice(1);

    const [_, setSpecialty] = useState(new Set([defaultSpecialty]));

    const params = useMemo(
      () => new URLSearchParams(searchParams),
      [searchParams],
    );

    const search = searchParams.get('search') ?? '';

    const handleReplaceURL = useCallback(
      (params: URLSearchParams) => {
        startTransition?.(() => {
          replace(`${pathname}?${params}`);
        });
      },
      [pathname, replace, startTransition],
    );

    const handlePageChange = useCallback(
      (page: number) => {
        page === 1
          ? params.delete('page')
          : params.set('page', page.toString());

        handleReplaceURL(params);
      },
      [handleReplaceURL, params],
    );

    const updateSearchParams = useCallback(
      (value: string) => {
        const specialty = searchParams.get('specialty');

        if (!specialty) {
          params.append('specialty', value);
        } else if (value) {
          params.set('specialty', value);
        } else {
          params.delete('specialty');
        }

        handleReplaceURL?.(params);
      },
      [handleReplaceURL, params, searchParams],
    );

    const handleSelectSpecialty = useCallback(
      (key: Key) => {
        setSpecialty(new Set([key as string]));

        // Find the selected specialty from specialtyOptions
        const selectedSpecialty = specialtyOptions.find(
          (option) => option.key === key,
        );

        if (key === 'all') {
          params.delete('specialty');
          handleReplaceURL?.(params);

          return;
        }

        const { label = '' } = selectedSpecialty || {};
        updateSearchParams(formatString(label));
      },
      [handleReplaceURL, params, specialtyOptions, updateSearchParams],
    );

    const isAdmin = role === ROLE.ADMIN;
    // Handle create
    const handleCreate = useCallback(() => {
      setChemist(undefined);
      setChemistId('');
      onOpen();
    }, [onOpen]);

    // Handle edit
    const handleEdit = useCallback(
      ({ data, id }: { data: UserModel; id: string }) => {
        if (!isAdmin) return;
        setChemist(data);
        setChemistId(id);
        onOpen();
      },
      [isAdmin, onOpen],
    );

    return (
      <>
        <div className="flex flex-col mt-3 md:flex-row gap-4 md:mb-8">
          <InputSearch placeholder="Search Chemists" value={search} />
          <div className="flex justify-between md:gap-4 mb-10 md:mb-0 ">
            <MenuDropdown
              icon={<CategoryIcon customClass="w-4 h-4 md:w-6 md:h-6" />}
              label="Specialty"
              options={specialtyOptions}
              classNames={{
                trigger: 'w-[120px] md:w-[170px] h-[52px]',
              }}
              selectionMode="multiple"
              onAction={handleSelectSpecialty}
            />
            {isAdmin && (
              <Button className="font-medium h-[52px]" onClick={handleCreate}>
                Create
              </Button>
            )}
          </div>
        </div>
        <Text variant="primary" size="2xl" customClass="my-5">
          {params.size ? 'Chemist Results' : 'All Chemists'}
        </Text>
        {isPending ? (
          <ChemistListSkeleton />
        ) : (
          <div className="flex flex-col items-center min-h-[40vh] sm:min-h-[60vh]">
            {chemists.length > 0 ? (
              <div className="w-full grid gap-8 grid-cols-1 md:grid-cols-2 2xl:grid-cols-3 min-[2048px]:grid-cols-4 justify-evenly justify-items-center">
                {chemists.map((chemist) => {
                  const { attributes } = chemist;
                  const { users_permissions_user } = attributes || {};
                  const { id = '', attributes: data } =
                    users_permissions_user?.data || {};

                  return (
                    <ChemistCard
                      id={id}
                      key={id}
                      isAdmin={isAdmin}
                      data={data}
                      onEdit={handleEdit}
                    />
                  );
                })}
              </div>
            ) : (
              <Text size="lg" variant="description" customClass="my-auto">
                {RESULT_NOT_FOUND}
              </Text>
            )}
            {!!pagination && pagination.pageCount > 1 && (
              <Suspense fallback={null}>
                <Pagination
                  classNames={{ base: 'mt-4' }}
                  initialPage={page}
                  total={pageCount}
                  onChange={handlePageChange}
                />
              </Suspense>
            )}
          </div>
        )}
        {isAdmin && (
          <ChemistModal
            isOpen={isOpen}
            id={chemistId}
            data={chemist}
            onClose={onClose}
            specialtyOptions={specialtyOptionsForm}
          />
        )}
      </>
    );
  },
);

ChemistList.displayName = 'ChemistList';
export default ChemistList;
