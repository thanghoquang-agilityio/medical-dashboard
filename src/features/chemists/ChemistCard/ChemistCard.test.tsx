import { render, screen } from '@testing-library/react';

// Components
import ChemistCard, { ChemistCardProps } from '.';

// Mock
import { MOCK_CHEMISTS_LIST } from '@/mocks/chemists';

describe('ChemistCard test cases', () => {
  const mockProps: ChemistCardProps = {
    data: MOCK_CHEMISTS_LIST[0].attributes.users_permissions_user.data
      .attributes,
    id: '1',
    isAdmin: true,
  };
  const setup = (props: ChemistCardProps) => render(<ChemistCard {...props} />);

  it('should render correctly', () => {
    const { asFragment } = setup(mockProps);

    expect(asFragment()).toMatchSnapshot();
  });

  // TODO: will update test in another MR
  it.skip('should display unknown when specialty is not defined', () => {
    setup({
      ...mockProps,
    });

    const message = screen.getByText(/unknown/i);

    expect(message).toBeInTheDocument();
  });

  // TODO: will update test in another MR
  it.skip('should display no description when there is no description', () => {
    setup({
      ...mockProps,
    });

    const message = screen.getByText(/no description/i);

    expect(message).toBeInTheDocument();
  });
});
