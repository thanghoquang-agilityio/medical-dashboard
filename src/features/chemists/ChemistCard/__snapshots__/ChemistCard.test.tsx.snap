// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChemistCard test cases should render correctly 1`] = `<DocumentFragment />`;
